proc/groupby.go

```diff
@@ -175,60 +179,82 @@ func (g *GroupBy) Pull() (zson.Batch, error) {
 	}
 }
 
-func (g *GroupByAggregator) createRow(ts nano.Ts, kvals [][]byte) *GroupByRow {
+func typeKey(rowkeys []zeek.TypedEncoding) string {
+	var b strings.Builder
+	for _, rowkey := range rowkeys {
+		b.WriteString(rowkey.Type.String())
+		b.WriteString(";")
+	}
+	return b.String()
+}
+
+func (g *GroupByAggregator) createRow(ts nano.Ts, rowkeys []zeek.TypedEncoding) *GroupByRow {
 	// Make a deep copy so the caller can reuse the underlying arrays.
-	kvals = append(make([][]byte, 0, len(kvals)), kvals...)
-	for k, v := range kvals {
-		if v != nil {
-			kvals[k] = append(make([]byte, 0, len(v)), v...)
-		}
+	v := make([]zeek.TypedEncoding, len(rowkeys))
+	copy(v, rowkeys)
+	key := typeKey(v)
+	_, ok := g.typeCols[key]
+	if !ok {
+		g.typeCols[key] = v
 	}
 	return &GroupByRow{
-		keyVals: kvals,
+		rowKeys: v,
 		ts:      ts,
 		columns: compile.Row{Defs: g.reducerDefs},
 	}
 }
 
+// XXX this could be made more efficient by using exporting zval.Encoding.build
+// and using it rowkey.Body.Build() and do something smarter with the zeek type
+// strings... otherwise we are sending lots of strings to the GC on each record
+// defeating the purpose of g.cacheKey.
+func (g *GroupByAggregator) key(key []byte, rowkeys []zeek.TypedEncoding) ([]byte, error) {
+	if len(rowkeys) > 0 {
+		key = append(key, rowkeys[0].Type.String()...)
+		for _, rowkey := range rowkeys[1:] {
+			key = append(key, ':')
+			key = append(key, rowkey.Type.String()...)
+		}
+		for _, rowkey := range rowkeys {
+			key = append(key, ':')
+			key = append(key, rowkey.String()...)
+		}
+	}
+	return key, nil
+}
+
 // Consume takes a record and adds it to the aggregation. Records
 // successively passed to Consume are expected to have timestamps in
 // monotonically increasing or decreasing order determined by g.reverse.
 func (g *GroupByAggregator) Consume(r *zson.Record) error {
 	// Extract the list of groupby expressions.  Re-use the array
 	// stored in consumeCutDest to avoid re-allocating on every record.
-	var vals [][]byte
-	if g.consumeCutDest != nil {
-		vals = g.consumeCutDest[:0]
-	}
-	for _, key := range(g.keys) {
-		_, v := key.resolver(r)
-		if v != nil {
-			vals = append(vals, v)
+	var rowkeys []zeek.TypedEncoding
+	if g.cacheRowKeys != nil {
+		rowkeys = g.cacheRowKeys[:0]
+	}
+	for _, key := range g.keys {
+		rowkey := key.resolver(r)
+		if rowkey.Body != nil {
```

Comment:
This should be `.Type` (this is PR #81 but don't want it to get lost in rebase)

Reply:
Obvious but subtle without thinking about it.  Maybe we should add a method on TypedEncoding for this check?